component/discovery/kubernetes_crds/kubernetes_crds.go

```diff
@@ -0,0 +1,336 @@
+			delete(c.discoveryConfigs, k)
+			delete(c.scrapeConfigs, k)
+		}
+	}
+}
+
+func (c *crdManager) addPodMonitor(pm *v1.PodMonitor) {
+	for i, ep := range pm.Spec.PodMetricsEndpoints {
+		pmc := c.cg.generatePodMonitorConfig(pm, ep, i)
+		c.mut.Lock()
+		c.discoveryConfigs[pmc.JobName] = pmc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[pmc.JobName] = pmc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+
+func (c *crdManager) addServiceMonitor(sm *v1.ServiceMonitor) {
+	for i, ep := range sm.Spec.Endpoints {
+		smc := c.cg.generateServiceMonitorConfig(sm, ep, i)
+		c.mut.Lock()
+		c.discoveryConfigs[smc.JobName] = smc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[smc.JobName] = smc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+
+func (c *crdManager) addProbe(p *v1.Probe) {
+	probe := c.cg.generateProbeConfig(p)
+	c.mut.Lock()
+	c.discoveryConfigs[probe.JobName] = probe.ServiceDiscoveryConfigs
+	c.scrapeConfigs[probe.JobName] = probe
+	c.mut.Unlock()
+	c.apply()
+}
+
+func (c *crdManager) onAddPodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	level.Info(c.logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.addPodMonitor(pm)
+}
+
+func (c *crdManager) onUpdatePodMonitor(oldObj, newObj interface{}) {
+	pm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.addPodMonitor(newObj.(*v1.PodMonitor))
+}
+func (c *crdManager) onDeletePodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.apply()
+}
+
+func (c *crdManager) onAddServiceMonitor(obj interface{}) {
+	sm := obj.(*v1.ServiceMonitor)
+	level.Info(c.logger).Log("msg", "found service monitor", "name", sm.Name)
+	c.addServiceMonitor(sm)
+}
+func (c *crdManager) onUpdateServiceMonitor(oldObj, newObj interface{}) {
+	sm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("serviceMonitor", sm.Namespace, sm.Name)
+	c.addServiceMonitor(newObj.(*v1.ServiceMonitor))
+}
+func (c *crdManager) onDeleteServiceMonitor(obj interface{}) {
+	sm := obj.(*v1.ServiceMonitor)
+	c.clearConfigs("serviceMonitor", sm.Namespace, sm.Name)
+	c.apply()
+}
+
+func (c *crdManager) onAddProbe(obj interface{}) {
+	p := obj.(*v1.Probe)
+	level.Info(c.logger).Log("msg", "found probe", "name", p.Name)
+	c.addProbe(p)
+}
+func (c *crdManager) onUpdateProbe(oldObj, newObj interface{}) {
+	p := oldObj.(*v1.Probe)
+	c.clearConfigs("probe", p.Namespace, p.Name)
+	c.addProbe(newObj.(*v1.Probe))
+}
+func (c *crdManager) onDeleteProbe(obj interface{}) {
+	p := obj.(*v1.Probe)
+	c.clearConfigs("probe", p.Namespace, p.Name)
+	c.apply()
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	// innerCtx gets passed to things we create, so we can restart everything anytime we get an update.
+	// Ideally, this component has very little dynamic config, and won't have frequent updates.
+	var innerCtx context.Context
+	// cancel is the func we use to trigger a stop to all downstream processors we create
+	var cancel func()
```

Comment:
I'd recommend calling cancel in a defer to be a little safer around ensuring it always gets canceled at the end: 

```go 
var cancel func() 
defer func() {
  if cancel != nil {
    cancel() 
  }
}()
```

Reply:
I think we're covered if the parent context cancels, but I have no problem being explicit for safety.